### PR TITLE
Fix split enclosures

### LINKED ISSUE #10: Separators inside special glob enclosures
Originally from https://github.com/es128/path-dirname/issues/1

``` js
> gp('/foo/{,/,bar/baz,qux}/')
'/foo/{,/,bar'
> gp('/foo[a\\\/]/')
'/foo[a\\'
```

is-glob used to return `true` for the presence of `{` or `[`, but it no longer does without the accompanying closing character.

Seems we need to do an extra truncation step over here for any unescaped sets of enclosure characters that might have a path separator inside. Is that only square and curly brackets, or are parentheses an issue as well?

### PR DESCRIPTION
**Pull request by es128**
Here's what I've got so far to resolve #10

@jonschlinkert if you've got more interesting test cases to throw at this to try to break it then bring them on

### CODE REVIEW

**Comment by phated** [comment]
Looks good to me if those sequences can't be placed in `( )`

**Comment by es128** [comment]
I'm going by https://github.com/jonschlinkert/is-extglob in my assumption that there is no scenario where `/` can legally be placed inside `()`

**Comment by es128** [comment]
Although maybe that's a false assumption based on the simple examples there. I'm not sure how sophisticated they're allowed to get.

**Comment by jonschlinkert** [comment]
looking at this now

File: index.js [hunk 1]
```diff
@@ -5,7 +5,16 @@ var isglob = require('is-glob');
 var pathDirname = require('path-dirname');
 
 module.exports = function globParent(str) {
-	str += 'a'; // preserves full path in case of trailing path separator
+	// replace braces/brackets sections with *
+	str = str.replace(/(^|[^\\])(\{([^{}]*?)}|\[([^\[\]]*?)\])/g, '$1*');
+	if (/[\}\}]$/.test(str)) str += '/';
```
Comment (jonschlinkert) [inline index.js]: repeated characters in a regex character class have no significance. e.g. `[\}\}]` is the same as `[\}]`

Comment (jonschlinkert) [inline index.js]: is the goal to basically trick is-glob into returning true for half of a brace/bracket/parens pattern?

Comment (es128) [inline index.js]: Oops one of those was supposed to be a closing square bracket

Comment (es128) [inline index.js]: This line is specifically meant to fix https://github.com/es128/glob-parent/pull/11/files#diff-1dd241c4cd3fd1dd89c570cee98b79ddR78

Where the special chars were escaped, but then still being split by dirname

Comment (jonschlinkert) [inline index.js]: Yeah, this won't work, none of the implementations require both braces to be escaped for the brace pattern to be escaped. Also, this fails on nested braces when only one of them is escaped.

Comment (es128) [inline index.js]: This only requires the first one to be escaped, but yeah I didn't look at any nested cases.

Think I should abandon this approach, or can it possibly be tweaked to do the job?

Either way seems I should add some nested braces cases to the tests.

**Comment by jonschlinkert** [comment]
Oh, I have another suggestion. @phated I thought about this before and forgot about it. IMHO if we were to do this "the right way", we would expand braces in glob patterns in glob-stream as the very first thing. Basically just take the array of given patterns and expand them all into a reduced array of patterns, free of braces. This is the first thing that every glob matcher does anyway. Once that's done, any braces left in the pattern would not be part of a brace set (e.g. they would be literal path characters - which, people do use sometimes). 

This mostly resolves the issue of having to figure out which braces are escaped etc. and I have to think it would result in more accurate glob operations too, since the `base` could be created per pattern (but that's a different issue).

Also, I mentioned this [here](https://github.com/es128/glob-parent/pull/11#discussion_r83977862), but more and more I think we should add an option to relax the matching in is-glob, so that it doesn't care if brackets, parens or braces are balanced. 

In cases where it's wrong, it would only be including extra path segments in the glob, but at least we wouldn't "clobber the glob" to coin a phrase.

**Comment by es128** [comment]
Sorry I stopped to have dinner as well. 

I totally appreciate your expertise on this, I was just attempting a quick/dirty solution in case it could have worked out to be good enough. 

If you'd rather implement a loose mode in is-glob I'd be happy to chuck this PR and use that instead.

**Comment by es128** [comment]
Don't want to go all the way back to the original simple regex though. Hopefully we can keep the improvement of knowing the difference between glob-relevant parens and dirnames with parens in them.

**Comment by jonschlinkert** [comment]
> knowing the difference between glob-relevant parens and dirnames with parens in them.

yeah I think we'd still want to do that here. I'm specifically concerned with the braces. it's impossible to do them correctly without using a real parser. I'm about to push up the refactor of braces i've been working on, I'll ping everyone when that's up

**Comment by es128** [comment]
@jonschlinkert I took another crack at it and updated the PR (I know, I know.. sorry)

It's handling all the scenarios you've brought to my attention fairly well, with just a few caveats

1. If only the closing brace/bracket is escaped it will truncate that section. I get that this doesn't quite meet the spec, but I think it's acceptable nonetheless because a) who does that?; b) the output is still reasonable: `gp('path/{../,./,{bar,/baz\\},qux\\}/'); //=> 'path'. Leading escapes are handled well, and I think more in line with likely user behavior in this sort of corner case: `gp('path/\\{,/,bar/baz,qux}/'); //=> 'path/{,/,bar/baz,qux}'`
2. Nested braces where the outside is escaped and the inner one is not currently fails badly (`gp('path/\\{../,./,{bar,/baz},qux}/'); //=> 'path/{../,.'`). But this is such a weird case that I don't mind accepting it as a limitation for the sake of keeping the implementation here light and simple.

We can call out the caveats in the readme. Do you think this would be acceptable? Do you have more scenarios you're worried about that haven't yet been discussed?

**Comment by jonschlinkert** [comment]
> I get that this doesn't quite meet the spec, but I think it's acceptable nonetheless because a) who does that?

lol, I totally understand where you're coming from. 

It just seems like we should use existing tools that already solve this problem the right way.  glob-parent and path-dirname would still be differentiated and necessary, but the glob, bracket and brace logic doesn't seem to belong here. I've already mentioned a couple of times that I think you might be underestimating the edge cases that come with these patterns. but maybe I'm overthinking it.

Braces are the biggest problem. To be reliably correct in a "defensible" way (e.g. "it's done this way because that's the spec") then we really should be expanding braces first. I don't know why I didn't think to do that a long time ago on all of our libs. It's the first thing done by all of the globbing implementations anyway (braces, minimatch, micromatch, etc), it's always completely separate from all other parsing/compiling operations, and brace expansion is always skipped when valid braces aren't detected in the pattern, so by expanding braces first I'm betting the net processing time will be faster.

**Comment by es128** [comment]
It just sounds heavy in terms of deps and processing to solve for what I'd expect to be some really rare cases. This used to be like 3 lines of simple code and one dep that was just a regex, and the only notable complaints I can think of have had to do with parentheses in dir names.

I'm fine with calling out the known limitations and leaving it to the small number of users in these odd situations (possibly zero) to find alternate solutions rather than impose a bunch of heavy lifting on everyone else to solve what used to be a simple concern for them.

If you have ideas for solutions that are faster/lighter than what I have here and solve for more cases then I'm happy to adopt them, but I'm having a hard time envisioning that being the case with the sort of direction you're describing.

**Comment by jonschlinkert** [comment]
- I think this should still be a "3 line lib", it shouldn't have to do parsing, which it's effectively doing now 
- is-glob can take options. having a "relaxed" option does not go against the purpose of that module IMO

I'm not sure where the disconnect is on this. I think this is a really simple suggestion that is way lighter than what you're doing. 

Now, _over and above that_, I also suggested expanding braces first in glob-stream so that didn't have to effectively parse the glob to figure out escape patterns (which is definitely the "heaviest" part of this operation besides hitting the file system in node-glob)

I think this makes sense for a couple of reasons:

- If braces are expanded first in glob-stream, we might need a lib for it, but the code wouldn't be brittle, we would follow established specs,   
- we wouldn't really be adding another step, we'd only be changing the order in which they happen. Since braces are always expanded first in globbing implementations (and brace expansion is a completely different parsing operation in all implementations I'm aware of. meaning it's not just done while parsing the glob, it's called separately), we would just do it first, then node-glob would detect the absence of braces and skip that.

(fwiw the refactor of braces is 100% bash compliant, passes 900 unit tests - including some that bash fails, per the notes in bash tests, and is very fast)

**Comment by es128** [comment]
I may have misinterpreted some of your statements to mean that you wanted to do heavier lifting here - and maybe you were talking about solutions applicable to the wider ecosystem, not just this lib.

If you think we can solve this effectively by adding a non-strict mode in is-glob and the change over here is limited to just enabling it then I'm totally cool with that.

**Comment by phated** [comment]
I think I'm on-board with the suggested changes but I wanted to clarify that I'd have to add another dependency to `glob-stream`?  I have been trying to reduce the dependencies in that library (hence the changes I've been landing in node-glob).

**Comment by jonschlinkert** [comment]
> I may have misinterpreted some of your statements to mean that you wanted to do heavier lifting here

no worries, I think we're on the same page now - I was a bit wordy, so probably my fault.  

> but I wanted to clarify that I'd have to add another dependency to glob-stream? I have been trying to reduce the dependencies in that library (hence the changes I've been landing in node-glob).

Correct. I knew you wanted to to reduce deps, so this is something I wanted to avoid.  But I think it might be good for the reasons stated, as well as a couple of other reasons that will be easier to explain here in a little bit when I push up some code and examples. 

That said, I'm looking at this like anything else: we're just trying to find the best solution, so I'm not making any assumptions that this is the best solution until we have a chance to see how the code comes together. And even then, you need to be comfortable with it. Anyway, I'm probably minutes away from pushing up braces, I'll explain more when that's up

**Comment by phated** [comment]
Sounds good.  I'm fine with adding dependencies if it produces the best result but I just wanted to clarify that I'm being very selective.

**Comment by phated** [comment]
@jonschlinkert was there any follow up to this that I needed to do?

**Comment by jonschlinkert** [comment]
sorry for the delay, I know you're trying to get this done and I'm the hold up.  I wanted to try to do some integration testing before we went any further with this. I have to wrap up another commitment, I should be able to review this tomorrow or this weekend at the very latest.

**Comment by phated** [comment]
@jonschlinkert no worries, you aren't a hold up.  I'm waiting on some node-glob stuff also.  I just thought I'd follow up because I think I have to add some code for brace expansion (this will go in a different feature than what I'm currently working on).

**Comment by jonschlinkert** [comment]
> I think I have to add some code for brace expansion

I was planning on doing a pr for that if that's ok with you. Totally your call, I don't mind doing it and adding tests around it.

**Comment by es128** [comment]
@jonschlinkert I'm not totally following how expanding braces first in glob-stream solves the issue in glob-parent for other users. Is the intention to document the limitation and instruct all users that they also need to expand braces prior to using glob-parent to avoid these potential problems?

**Comment by phated** [comment]
@jonschlinkert works perfectly for me!

@es128 I think the idea is to document that because it is an advanced feature

**Comment by es128** [comment]
> I think the idea is to document that because it is an advanced feature

https://github.com/jonschlinkert/glob-base/issues/3#issuecomment-256928428

So, the patch in this PR isn't perfect, but I think it drastically cuts down on the likelihood of people bumping into the remaining corner cases with some fairly simple additions to the code.

Rather than just saying "buyer beware" and forcing all downstream libs to do more work if they want to upgrade and do better with the substantially problematic parentheses situation while not stopping supporting some other useful glob paradigms, I think we should still use this or something like it regardless of whether glob-stream will now sidestep the concern by expanding braces before getting here. It doesn't have to be either/or.

@jonschlinkert sorry if I'm pestering, but considering what I just saw on glob-base, it seemed worth it to try to make the case again. Let me know what you think.

**Comment by jonschlinkert** [comment]
I don't feel like you're pestering at all. IMHO, when we're dealing with modules that are depended on as much as these are - and we want them to be stable and easy to maintain - it's worthwhile to discuss as much as we need to get it right. 

> Rather than just saying "buyer beware" and forcing all downstream libs to do more work if they want to upgrade and do better with the substantially problematic parentheses situation while not stopping supporting some other useful glob paradigms

I think it depends on how we look at this. I personally don't think you should even try to handle the corner cases here. Using `is-glob` to detect globs makes total sense, and if is-glob doesn't do what's needed to get the glob parent, then perhaps that's either a bug in is-glob, or we're overcomplicating what's necessary for getting the glob parent, or a mix of the two. I think the most important question to answer here is: how accurate does the glob parent really need to be? We know that having the glob parent is important for a couple of reasons, but does it really matter if it's one directory up (because we lop of an extra segment when a glob character exists, even if it's escaped)?

> but considering what I just saw on glob-base, it seemed worth it to try to make the case again. Let me know what you think.

I haven't had time to re-look at glob-base to decide what to do with it, but I'm pretty sure I'm going to deprecate it anyway after I publish the refactor of micromatch. Now that I'm using "real" parsers in all of the matching libs, glob base is no longer needed (or accurate enough). 

That said, that issue you're referring to on glob-base seems expected. He bumped dependencies with major breaking changes, but didn't update the tests or code in glob-base accordingly. Of course the tests will fail lol

**Comment by es128** [comment]
> but didn't update the tests or code in glob-base accordingly

Well, specifically it just re-exposed the problem of separators inside braces because you had that in the tests there. Nothing wrong with the tests, it's just a situation glob-parent does not handle in its current form.

It sounds to me at this point like what you're advocating for is effectively:

```diff
diff --git a/index.js b/index.js
index 9a7e593..b1a5e98 100644
--- a/index.js
+++ b/index.js
@@ -6,6 +6,7 @@ var pathDirname = require('path-dirname');

 module.exports = function globParent(str) {
        str += 'a'; // preserves full path in case of trailing path separator
-       do {str = pathDirname.posix(str)} while (isglob(str));
+       do {str = pathDirname.posix(str)}
+       while (isglob(str) || /[\{\[]/.test(str));
        return str.replace(/\\([\*\?\|\[\]\(\)\{\}])/g, '$1');
 };
```
(or an equivalent where the exact same thing happens inside is-glob based on an option)

i.e. dial down the impact of the is-glob improvements a bit and make it behave similarly to how it did before with the simple regex.

Not that I'm overly attached to my patch here, but I do feel like it achieves a reasonable compromise of addressing almost all of the cases of concern with hardly any added complexity. I'd be fine with the simpler approach too, but wrt this slightly more sophisticated PR as it is I'm still left feeling like I don't have a good answer to "why not?"

**Comment by jonschlinkert** [comment]
> I'd be fine with the simpler approach too, but wrt this slightly more sophisticated PR as it is I'm still left feeling like I don't have a good answer to "why not?"

To be clear, I do agree with your opinion that it's totally fine to leave in, and it's going to smooth out edge cases. But it also should be clear that your proposed patch will only be "wrong less often", since it can't possibly check for all of the edge cases. and again, I'm not saying that's bad. I asked earlier "how accurate do we really need to be?" and I think we need to answer that to decide how much this matters, because there are solutions that we can use to be as close as we're ever going to get to 100% accurate. It just depends on the priorities.

**Comment by jonschlinkert** [comment]
> I'm still left feeling like I don't have a good answer to "why not?"

I forgot to answer this part. even though glob-parent is supposed to return the directory only, and no glob parts, I'm still uneasy about the idea of unescaping the regex characters. Maybe the user escaped those characters just for glob-parent, maybe not. It just seems like it creates the potential for those characters to be read incorrectly upstream (if unescaped). 

For example, this would now make it harder to get the glob "base" from a path (e.g. get the pattern, stripped of path segments), because you wouldn't just be able to strip `foo/{bar}/` from `foo/\\{bar}/**`, because the `\\` is missing, so the paths wouldn't match up anymore. That's the first one that comes to mind.

**Comment by es128** [comment]
> it also should be clear that your proposed patch will only be "wrong less often"

I agree, and I've said as much in my comments after the last update to the PR. I think it's a reasonable balance between supporting every case I'm aware of that I'm concerned with supporting without bloating the lib - while documenting the known caveats.

> this would now make it harder to get the glob "base" from a path

I don't understand the use-case where it makes sense to preserve the escape characters in the output. The assertion is that what comes out of glob-parent should be an actual dir on the disk, so there's no more reason to run it through is-glob, or extract a pattern from it, or anything like that. As I've said before regarding the unescaping, I want to be able to pass the glob-parent output straight to `fs.readdir` or stuff like `fs.watch`, which won't work if the escape characters are still there.

If you're talking about it being difficult to use this to do what glob-base does - getting back the part of the pattern from where it gets globby, separate from the parent - then imo that's not the core use case and something that we should be able to figure out how to resolve from within glob-base. Maybe the compromise would be to offer a `preserveEscapes: true` option over here - if you intend to keep supporting glob-base.

**Comment by es128** [comment]
By the way, sorry if you've answered this and I've missed it - is it legal in globs to ever have a separator enclosed in parentheses? 

`/path/(some/subdirs|other/subdirs)`

**Comment by jonschlinkert** [comment]
> is it legal in globs to ever have a separator enclosed in parentheses?

Not in Bash (extglob), and I'm not sure about minimatch (I'm pretty sure it's not "disallowed" in minimatch, but I'm not sure. it might say on the readme), but it is allowed in extglob (my version), braces, and micromatch. 

IMHO, it should be supported, and the only reasons I can think of that Bash doesn't support it (and/or minimatch) is due to limitations of the parser and/or because of potential conflicts with other Bash expansions.

With braces (and micromatch), for example, `a/{b,c/d}/e` expands to `a/(b|c\/d)/e`, whereas minimatch would expand that to multiple patterns, then join them with `|` (so that path wouldn't be in parens). But again, I think that's a limitation, not a standard or convention. 

> there's no more reason to run it through is-glob, or extract a pattern from it, or anything like that

No? Why is something like this not a good approach? 

```js
var path = require('path');
var parent = require('glob-parent');
var pattern = 'node_modules/*/*.js';
var dir = parent(pattern);
var glob = require('glob');
var base = pattern.replace(new RegExp(dir + '\\/?'), '');
console.log(glob.sync(base, {cwd: dir}));
```

This returns the same relative paths that glob-stream would return.

**Comment by es128** [comment]
I was referring to the parent itself. It's not a pattern anymore, so the use-case you're showing with `pattern.replace(new RegExp(dir + '\\/?'), '')` is the only reason to care about where the escape characters were anymore. I think ultimately that code needs a better approach that deals with the possibility of escape characters in the input that are not in the output.

Won't the `{cwd: dir}` part in the last line break because now `dir` might not actually be the dir path? It can have escape characters left in it. I just verified that `fs.readdir` does not deal with the escape characters being there, they have to be removed.

I think the core use case is having the actual path to the dir.

**Comment by jonschlinkert** [comment]
hey would you guys want to hop on a call for a few min? I think I can explain this better in a convo. if not, no worries I can try to explain more here

**Comment by phated** [comment]
I can jump on a call now.

**Comment by jonschlinkert** [comment]
can you email me your mobile?

**Comment by es128** [comment]
@jonschlinkert @phated 

Sorry for the long delay, but this is ready to be reviewed again now.

**Comment by phated** [comment]
👍  LGTM

**Comment by jonschlinkert** [comment]
sorry guys, fighting off a bad cold. I'll go with your judgement on this. Last time I reviewed it looked good

**Comment by es128** [comment]
feel better soon @jonschlinkert 

Will go ahead and merge this, and I've got another PR coming to help windows/backslash users with simple use-cases such as in #12.
